File: app/routers/scanner.py
from fastapi import APIRouter, Security, Depends, Query
from app.schemas.scanner import ScanningStatistics, ScannerSettingsUpdate, FilterHistoricPings
from app.crud.scanner import get_settings, update_settings, get_stats, get_blocklist_fnames, get_historic_pings
from app.crud.users import get_current_user
from app.models.user import User
from typing import Annotated
from datetime import datetime
from typing import Optional

router = APIRouter()

@router.get("/settings/", tags=["settings"])
def fetch_settings(current_user: Annotated[User, Security(get_current_user, scopes=["admin"])]):
    return get_settings()

@router.post("/settings/", tags=["settings"])
def change_settings(update: ScannerSettingsUpdate, current_user: Annotated[User, Security(get_current_user, scopes=["admin"])]):
    update_settings(update)
    return {"status": "success"}

@router.get("/settings/available_blacklist_files", tags=["settings"])
def get_available_blacklist_files(current_user: Annotated[User, Depends(get_current_user)]):
    return get_blocklist_fnames()

@router.get("/statistics/", tags=["settings"])
def get_statistics(current_user: Annotated[User, Depends(get_current_user)]):
    return get_stats()

@router.get("/historic_pings/", tags=["historic_pings"])
def fetch_historic_pings(
    current_user: Annotated[User, Depends(get_current_user)],
    uri: Optional[str] = Query(None),
    raw_headers_keyword: Optional[str] = Query(None),
    speed_min: Optional[float] = Query(None),
    speed_max: Optional[float] = Query(None),
    error_type: Optional[str] = Query(None),
    start_date: Optional[datetime] = Query(None),
    end_date: Optional[datetime] = Query(None),
    limit: int = Query(50, ge=1, le=500)
):
    """
    Fetch historic ping records with optional filters.
    
    Filters:
    - uri: exact URI match
    - raw_headers_keyword: case-insensitive regex search in headers
    - speed_min/speed_max: filter by speed range
    - error_type: filter by error type
    - start_date/end_date: filter by ping date range
    - limit: max results (1-500, default 50)
    """
    filter_obj = FilterHistoricPings(
        uri=uri,
        raw_headers_keyword=raw_headers_keyword,
        speed_min=speed_min,
        speed_max=speed_max,
        error_type=error_type,
        start_date=start_date,
        end_date=end_date,
        limit=limit
    )
    return get_historic_pings(filter_obj)

File: frontend/src/api/scanner.js
import { getAuthHeader } from './user';
const API_BASE = '/api';
export async function getSettings() {
    const response = await fetch(`${API_BASE}/scanner/settings/`, {
        headers: getAuthHeader(),
    });
    if (!response.ok) {
        throw new Error('Failed to fetch scanner settings');
    }
    return response.json();
}
export async function updateSettings(settings) {
    const response = await fetch(`${API_BASE}/scanner/settings/`, {
        method: 'POST',
        headers: {
            ...getAuthHeader(),
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(settings),
    });
    if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to update scanner settings');
    }
}
export async function getStatistics() {
    const response = await fetch(`${API_BASE}/scanner/statistics/`, {
        headers: getAuthHeader(),
    });
    if (!response.ok) {
        throw new Error('Failed to fetch scanner statistics');
    }
    return response.json();
}
export async function getAvailableBlacklistFiles() {
    const response = await fetch(`${API_BASE}/scanner/settings/available_blacklist_files`, {
        headers: getAuthHeader(),
    });
    if (!response.ok) {
        throw new Error('Failed to fetch available blacklist files');
    }
    return response.json();
}
export async function getHistoricPings(filters) {
    const params = new URLSearchParams();
    if (filters) {
        if (filters.uri) {
            params.append('uri', filters.uri);
        }
        if (filters.raw_headers_keyword) {
            params.append('raw_headers_keyword', filters.raw_headers_keyword);
        }
        if (filters.speed_min !== undefined) {
            params.append('speed_min', String(filters.speed_min));
        }
        if (filters.speed_max !== undefined) {
            params.append('speed_max', String(filters.speed_max));
        }
        if (filters.error_type) {
            params.append('error_type', filters.error_type);
        }
        if (filters.start_date) {
            params.append('start_date', filters.start_date);
        }
        if (filters.end_date) {
            params.append('end_date', filters.end_date);
        }
        if (filters.limit !== undefined) {
            params.append('limit', String(filters.limit));
        }
    }
    const query = params.toString();
    const url = query ? `${API_BASE}/scanner/historic_pings/?${query}` : `${API_BASE}/scanner/historic_pings/?limit=500`;
    const response = await fetch(url, {
        headers: getAuthHeader(),
    });
    if (!response.ok) {
        throw new Error('Failed to fetch historic pings');
    }
    return response.json();
}
